cuda/matrix/dense_kernels.cu

```diff
@@ -332,8 +332,12 @@ void compute_norm2(std::shared_ptr<const CudaExecutor> exec,
 {
     if (cublas::is_supported<ValueType>::value) {
         auto handle = cublas::init();
-        cublas::norm2(handle, result->get_size()[1], result->get_values(),
-                      result->get_stride(), result->get_values());
+        for (size_type col = 0; col < x->get_size()[1]; ++col) {
+            cublas::norm2(handle, x->get_size()[0], x->get_const_values() + col,
+                          x->get_stride(),
+                          reinterpret_cast<remove_complex<ValueType> *>(
+                              result->get_values() + col));
```

Comment:
How would you put it in the interface? I need to call a kernel in order to do that properly, or is there some other way?

Reply:
Exactly. You use a different macro in `cuda/base/cublas_bindings.hpp` for the complex case, which also calls the additional kernel. Something like:

```c++
#define BIND_CUBLAS_NORM2(ValueType, ResultType, CublasName)                   \
    inline void norm2(cublasHandle_t handle, int n, const ValueType *x,        \
                      int incx, ResultType *result)                            \
    {                                                                          \
        ASSERT_NO_CUBLAS_ERRORS(CublasName(handle, n, as_culibs_type(x), incx, \
                                           as_culibs_type(result)));           \
    }

#define BIND_CUBLAS_COMPLEX_NORM2(ValueType, ResultType, CublasName)                   \
    inline void norm2(cublasHandle_t handle, int n, const ValueType *x,        \
                      int incx, ResultType *result)                            \
    {                                                                          \
        ASSERT_NO_CUBLAS_ERRORS(CublasName(handle, n, as_culibs_type(x), incx, \
                                           as_culibs_type(reinterpret_cast<...>(result))));           \
        <your kernel here> \
    }

BIND_CUBLAS_NORM2(float, float, cublasSnrm2);
BIND_CUBLAS_NORM2(double, double, cublasDnrm2);
BIND_CUBLAS_COMPLEX_NORM2(std::complex<float>, std::complex<float>, cublasScnrm2);
BIND_CUBLAS_COMPLEX_NORM2(std::complex<double>, std::complex<double>, cublasDznrm2);
```

Reply:
This should be now done. I am unsure where to put the interface and kernel for zeroing the array as the model doesn't quite fit the components one (`.cuh` files providing purely kernels) whereas here I need a combination of `.hpp` and `.cu` files.

Reply:
One option is to just change the extension to `.cuh` to make it consistent with the naming (it has no other consequences, since `.cuh` is a purely made up extension) and have the kernel defined in the header. With this option though, we need to compile it multiple times (one for each compilation unit that includes the header), and the header can only be used in `.cu` files.

Another option would be to just declare a function that will invoke the kernel and define the function and the kernel in an associated `cublas_bindings.cu` implementation file. This allows us to keep the `.hpp` extension and call the header from non-CUDA code. (There's nothing preventing a `.cu` source to have an `.hpp` header if there's no CUDA code in it, but the opposite is not true.)

Reply:
But don't do things like having a source file `x.a` and declaring its public interface in header `y.b`. If you need something in `y.b`, either implement it in `y.a`, or if you want / need to implement it in `x.a`, then create a header `x.b` and include it into `y.b`.

Reply:
Well option 1) is impossible due to needing this header precisely outside of `cu` world.

I did option 2) already, but I guess I will move everything to the components nonetheless.